### PR TITLE
Authentication api

### PR DESCRIPTION
**Pull request by venkylads**
Complete Authentication is added like sign in, sign up, logout, refreshing tokens and forget password also implemented passport authentication. Please review my code and let me know if there are any changes need to be done.

### CODE REVIEW

**Comment by AKHIL-882** [comment]
Remove all the info() functions

File: online-dashboard-api/app/Http/Controllers/AuthenticationController.php [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+use Throwable;
+
+class AuthenticationController extends Controller
+{
+    public function signUp(SignupRequest $request)
+    {
+
+        try {
+            //creating a new user
+            $user = User::create([
```
Comment (AKHIL-882) [inline online-dashboard-api/app/Http/Controllers/AuthenticationController.php:25]: Move the creation into the model

File: online-dashboard-api/app/Http/Controllers/AuthenticationController.php [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+            info($user);
+
+            info($tokenData);
+
+            //check if token generation failed
+            if (! $tokenData) {
+                //delete the user if token generation fails
+                $user->delete();
+
+                return response()->json([
```
Comment (AKHIL-882) [inline online-dashboard-api/app/Http/Controllers/AuthenticationController.php:43]: Create a seperate HTTP file ApiResponse and pass the resonse to the function to render data!

File: online-dashboard-api/app/Http/Controllers/AuthenticationController.php [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+    {
+
+        $refreshToken = $request->refresh_token;
+
+        try {
+            // Create an instance of AccessTokenController to handle the request
+            $accessTokenController = app(AccessTokenController::class);
+
+            // Create a Psr17Factory instance to generate PSR-7 compatible request
+            $psr17Factory = new Psr17Factory;
```
Comment (AKHIL-882) [inline online-dashboard-api/app/Http/Controllers/AuthenticationController.php:168]: Move this entire server request to a seperate file and get the URL http://127.0.0.1:8000/oauth/token from the config

File: online-dashboard-api/app/Models/User.php [hunk 4]
```diff
@@ -45,4 +50,25 @@ protected function casts(): array
...
     }
+
+    public function passwordResetTokens(): HasMany
+    {
+        return $this->hasMany(ForgetPasswordTokens::class);
+    }
+
+    public function initiatePasswordReset(): void
+    {
+        $expiry = Carbon::now()->addMinutes(60);
```
Comment (AKHIL-882) [inline online-dashboard-api/app/Models/User.php:61]: Make expiry count to 5 mins and get the value from config tooo

**Review by AKHIL-882 — CHANGES_REQUESTED** [review]

**Comment by AKHIL-882** [comment]
move all the values (integers) to the config files and refactor the files into folders which fits in

File: online-dashboard-api/routes/api.php [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+Route::middleware(['auth:api', 'api'])->group(function () {
+
+    Route::match(['post', 'get'], '/logout', [AuthenticationController::class, 'logout']);
+
+    Route::post('refresh-token', [AuthenticationController::class, 'refreshAccessToken']);
+
+});
+
+Route::middleware(['guest', 'throttle:10,1'])->group(function () {
+
```
[the remark this answers is not in the document]
Comment (venkylads) [reply online-dashboard-api/routes/api.php:16]: routes like homepage, signup, login and reset password can be under the guest middleware to allow non-authenticated users to access these routes. routes like logout, refresh tokens should be under the auth:api middleware to protect the routes from non-authenticated users

File: online-dashboard-api/app/Http/Controllers/AuthenticationController.php (hunk 1 above)
Comment (venkylads) [reply online-dashboard-api/app/Http/Controllers/AuthenticationController.php]: done

File: online-dashboard-api/app/Http/Controllers/AuthenticationController.php (hunk 2 above)
Comment (venkylads) [reply online-dashboard-api/app/Http/Controllers/AuthenticationController.php]: done

File: online-dashboard-api/app/Models/User.php (hunk 4 above)
Comment (venkylads) [reply online-dashboard-api/app/Models/User.php]: done

File: online-dashboard-api/app/Http/Responses/ApiResponse.php [hunk 6]
```diff
@@ -0,0 +1,35 @@
+<?php
```
Comment (AKHIL-882) [inline online-dashboard-api/app/Http/Responses/ApiResponse.php:1]: Excellent work! 🔥

Can we write the entire file in this manner! 

`public function setMessage($message): self
    {
        $this->message = $message;

        return $this;
    }`
    
    In place of getTokens we can replace with 
    `    public function mergeResults(array $results = []): self
    {
        $this->response = array_merge($this->results, $results);
        
        return $this;
    }`
    
    
For responses you can write the different function based on the code
like
`    public function respondOK(): JsonResponse
    {
        return $this->setStatusCode(200)->respondWithSuccess();
    }`
    
    
and use them whereever you want!

**Pull request merged**